### ISSUE TITLE
Bug: [no-floating-promise] Intersection types are not recogised by `allowForKnownSafePromises`

### ISSUE THREAD

**Issue by TastyPi** [typescript-eslint/typescript-eslint#11907]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=PTAEBUAsEsGdTqALpApqAJqgZtAdtEtAPZ6jYBOxAtuQIaxHYCeAUKgB4AOxFSyzLugDKdbKgAKVanFQAZaAGtUAHnAA%2BUAF5QUmrIXK1mgGSgA3qAD6VgDb4kqCgCEKdPBlgAuUAHJR4noysPJKqL6gAL6srNgArngAxkSk5MTEABQAlD4BktIGYSoJWLh4qBia5qygoBSoSHEUZCU4%2BBWgDKAJinjEAO5kXXlBhUatZRXqrNGsIKAAYgxMzACSeIzuiagAdPUA5nCOFHUNTRudi8vQLOubSao7T6agIwUhhqoT7ZWxCckkMgAIzoFGyPnMkVAZje%2Bg%2BRW%2B5UqFhqp0azW6HjaSM68B6fUGuIsUJhYnycNC4yxk1%2Bs2w6WyMRBYKyQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTowAbUiJoE6JAA0mcNhz94fAO4AxDtADSnQ0wDK-bogAKE6bNQYFO3RGBRuL%2BUhuSiRNKCZ%2BSURAuwdnKRlEABlKAGtosABfL2wAXQ0cnDIAC0QyVIAVUoiAIyR3MHw4RBzM3J1szKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// This is the definition from fastify
export type SafePromiseLike<T> = PromiseLike<T> & { __linterBrands: 'SafePromiseLike' }

function foo(): SafePromiseLike<undefined> {
  return undefined as unknown as SafePromiseLike<undefined>
}

// FastifyInstance.register returns a FastifyInstance<...> & SafePromiseLike<undefined>
function bar(): {} & SafePromiseLike<undefined> {
  return undefined as unknown as {} & SafePromiseLike<undefined>
}

foo()

bar()
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-floating-promises": ["error", {
      "allowForKnownSafePromises": [
        { "from": "file", "name": "SafePromiseLike" }
      ],
      "checkThenables": true
    }]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Neither `foo()` nor `bar()` are complained about

### Actual Result

`bar()` is complained about

### Additional Info

This was supposed to have been fixed by #9303, I don't know if it did not actually fix the issue or if this is a regression.

**Comment by TastyPi** [comment]
I've been trying to enable `checkThenables` in a repo that uses `fastify`, and ran into this issue. Fastify created `SafePromiseLike` specifically for this use case, but it seems like it doesn't work.

`FastifyInstance.register(...)` returns a `FastifyInstance<...> & SafePromiseLike<undefined>`. If I do this:

```ts
{
  allowForKnownSafePromises: [
    { from: "package", name: "SafePromiseLike", package: "fastify" },
  ],
  checkThenables: true,
}
```

it still errors on `FastifyInstance.register(...)` calls.

However, if I do this:

```ts
{
  allowForKnownSafePromises: [
    { from: "package", name: "FastifyInstance", package: "fastify" },
  ],
  checkThenables: true,
}
```

This it does match the `FastifyInstance.register(...)` correctly and does not complain.

My only insight into why is `FastifyInstance` is an `interface` whereas `SafePromiseLike` is a `type`. I suspect `SafePromiseLike` is being expanded into its own intersection before the check.

**Comment by bradzacher** [comment]
This is happening due to a fun implementation detail of TS.

If you use the "Types" tab in the playground you can see what's happening.

When I click on the `Identifier` for the type declaration on line 2 you can see that it has `typeToString(): SafePromiseLike<T>` with ID `93` (in my playground).

The return type of `foo` is a `TypeReference` node whose return type is a `TypeReference` that has `typeToString(): SafePromiseLike<T>`. If you click into the `Identifier` of this type you can see TS is telling us that the resolved type has ID `93` (i.e. TS has resolved the name back to the type alias on line 2).

So our logic is able to match these two declarations together and avoid reporting as expected.

However the return type of `bar` is an `IntersectionType`. If you inspect just this type you can immediately see the problem. TS has unravelled the type. Instead of the `typeToString()` being `{} & SafePromiseLike<undefined>`, it is `PromiseLike<undefined> & { __linterBrands: "SafePromiseLike"; }`.

And to further confirm this -- if you inspect the `.types` of the `IntersectionType` you can see the first element is a type which resolves to `PromiseLike<undefined>` and the second element is a type which resolves to `{ __linterBrands: 'SafePromiseLike' }`.

So TS has effectively erased the presence of the name `SafePromiseLike` from the type. This means that our rule cannot in any way see that this intersection is related to the `SafePromiseLike` type -- which is why it reports.

There's nothing we can do about this -- this is just a limitation of TS. The limitation is that if you intersect and intersection then TS will flatten the resulting type into a brand new type.

Put another way:
```ts
type T = {t: string};
type U = {u: string};
type V = {v: string};

type TU = T & U;
//   ^? type TU = T & U
type TU_V = TU & V;
//   ^? type TU_V = T & U & V
type T_U_V = T & U & V;
//   ^? type T_U_V = T & U & V

// i.e. TU_V === T_U_V !== (TU & V)
```

<img width="281" height="210" alt="screenshot of playground" src="https://github.com/user-attachments/assets/064bfc54-20b0-472c-82b9-42fc055a614f" />
